RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,404 @@
+
+  //local parameters of upper cluster in lower system of reference
+  Local3DPoint lparamsUpperInLower = geomDetLower->surface().toLocal(gparamsUpper);
+
+  LogTrace("VectorHitBuilderAlgorithm") << "\t lower global pos: " << gparamsLower;
+  LogTrace("VectorHitBuilderAlgorithm") << "\t upper global pos: " << gparamsUpper;
+
+  LogTrace("VectorHitBuilderAlgorithm") << "A:\t lower local pos: " << lparamsLower.first
+                                        << " with error: " << lparamsLower.second << std::endl;
+  LogTrace("VectorHitBuilderAlgorithm") << "A:\t upper local pos in the lower sof " << lparamsUpperInLower
+                                        << " with error: " << lparamsUpper.second << std::endl;
+
+  bool ok =
+      checkClustersCompatibility(lparamsLower.first, lparamsUpper.first, lparamsLower.second, lparamsUpper.second);
+
+  if (ok) {
+    AlgebraicSymMatrix22 covMat2Dzx;
+    double chi22Dzx = 0.0;
+    Local3DPoint pos2Dzx;
+    Local3DVector dir2Dzx;
+    fit2Dzx(lparamsLower.first,
+            lparamsUpperInLower,
+            lparamsLower.second,
+            lparamsUpper.second,
+            pos2Dzx,
+            dir2Dzx,
+            covMat2Dzx,
+            chi22Dzx);
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  pos2Dzx: " << pos2Dzx;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  dir2Dzx: " << dir2Dzx;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  cov2Dzx: " << covMat2Dzx;
+    VectorHit2D vh2Dzx = VectorHit2D(pos2Dzx, dir2Dzx, covMat2Dzx, chi22Dzx);
+
+    AlgebraicSymMatrix22 covMat2Dzy;
+    double chi22Dzy = 0.0;
+    Local3DPoint pos2Dzy;
+    Local3DVector dir2Dzy;
+    fit2Dzy(lparamsLower.first,
+            lparamsUpperInLower,
+            lparamsLower.second,
+            lparamsUpper.second,
+            pos2Dzy,
+            dir2Dzy,
+            covMat2Dzy,
+            chi22Dzy);
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  pos2Dzy: " << pos2Dzy;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  dir2Dzy: " << dir2Dzy;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  cov2Dzy: " << covMat2Dzy;
+    VectorHit2D vh2Dzy = VectorHit2D(pos2Dzy, dir2Dzy, covMat2Dzy, chi22Dzy);
+
+    OmniClusterRef lowerOmni(lower);
+    OmniClusterRef upperOmni(upper);
+    VectorHit vh = VectorHit(*stack, vh2Dzx, vh2Dzy, lowerOmni, upperOmni);
+    return vh;
+  }
+
+  return VectorHit();
+}
+
+void VectorHitBuilderAlgorithm::fit2Dzx(const Local3DPoint lpCI,
+                                        const Local3DPoint lpCO,
+                                        const LocalError leCI,
+                                        const LocalError leCO,
+                                        Local3DPoint& pos,
+                                        Local3DVector& dir,
+                                        AlgebraicSymMatrix22& covMatrix,
+                                        double& chi2) const {
+  std::vector<float> x = {lpCI.z(), lpCO.z()};
```

Comment:
use local arrays!

Reply:
These vectors are later on passed to a LinearFitter that takes vectors are arguments: https://github.com/cms-sw/cmssw/blob/master/CommonTools/Statistics/interface/LinearFit.h#L16-L25 

As the vectors would have to be created anyway in the fit() function, I don't expect much improvement by first using arrays and converting them to vectors later.

Reply:
Thanks for that suggestion! I was able to use that to replace all the vectors with local arrays.

Reply:
Turns out this is, at least as implemented by me, very slow and apparently memory-leaky, on cmslpc cmsRun even dies because it runs out of memory. As I don't really see the arrays being the problem, I suspect the linearFit(), but it would be great if you could have a look at the current implementation of the arrays to see if I am overlooking something.

Reply:
It's not immediately obvious to me why the array version would have issues. Are you sure this is the change that caused the problem?

Reply:
Pretty much, went back to the commit before the problem started and then applied the changes one by one by hand. This one seemed to create the issue.

Reply:
Turns out that the templated version expects the square or the uncertainties as the input while the vector version expects just the uncertainties... This is resolved now and we can stay with the arrays.

Reply:
Aha, thanks for finding that difference.